### PR TITLE
KEYCLOAK-5130 Add cookie support for Spring Security adapter

### PR DESCRIPTION
**Pull request by scranen**
Notes to the reviewers:

- The first commit is actually a workaround for KEYCLOAK-4342 (see #4056). Have a look; I think it could be a pragmatic first improvement without needing to implement a reliable way of establishing the context path. If it's good enough I'll turn it into a separate PR.
- It feels like the cookie based login redirect (2nd commit) should be in a different place, but this is what I came up with to get it to work for our application. Maybe someone who knows a bit more about the internal state keeping of the adapters can suggest a nicer way of doing this.
- I had to change the position of the `KeycloakAuthenticationProcessingFilter` in the filter chain (3rd commit), because otherwise the logout would be initiated before the credentials were retrieved from the cookie. I'm unsure what the implications are for the SSO logout in relation to Keycloak sessions and HTTP sessions (if enabled), but it seems to do what I want in local testing.
- I have not spent any effort on writing tests yet, because I'm unsure where/how to add tests (is there an existing integration test I can amend?). Any pointers would be appreciated.
- I'm unsure if these changes affect the behaviour of the Spring Boot adapter. Looking at the code, it seems that adapter operates independently from the Spring Security one, but the `pom.xml` does declare `keycloak-spring-security-adapter` as a dependency, so maybe there's some class path magic going on that I did not spot.

### CODE REVIEW

**Comment by sebastienblanc** [comment]
I will try to free some time this week to review this. 
Concerning your first commit maybe it's better to directly create a new PR since it touches the core adapter and there might be more impact than we thought.

**Comment by scranen** [comment]
I created #4356 for this.

**Comment by sebastienblanc** [comment]
Ok, we should try to get this in asap, @scranen could you rebase (and the "workaround" PR) to check if everything is still in stable state ?

**Comment by scranen** [comment]
I rebased onto the KEYCLOAK-4342 branch. I will try to find time next week to do some testing.

**Comment by scranen** [comment]
Rebased again to resolve conflicts. The `KeycloakAuthenticationToken` constructor has gained an extra parameter, but it's not documented so I had to guess the value. Maybe @glavoie can have a look to check that I did it right?

**Comment by glavoie** [comment]
@scranen, I had a quick look and I think it looks good.

This extra parameter was added as a fix to KEYCLOAK-5499 (https://github.com/keycloak/keycloak/pull/4488). Previous code used the check on the authentication credential passed to the adapter as a way to discern non-interactive (pre-emptive) authentication from interactive (with a back&forth to Keycloak to display the login page). Some logic differs based on how you want to authenticate the user.

**Comment by stianst** [comment]
For ConcurrencyTest see https://issues.jboss.org/browse/KEYCLOAK-5617. It's not stable at the moment. You don't happen to have the log do you? If so could you drop it into the issue as a comment please?

File: adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/CookieTokenStoreTest.java
Comment (sebastienblanc) [inline adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/CookieTokenStoreTest.java]: Could you add the license header ?

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationSuccessHandler.java
Comment (sebastienblanc) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationSuccessHandler.java]: Could you add the license header ?

**Comment by sebastienblanc** [comment]
@trecloux Do you mind reviewing this ? ;)

**Comment by bramvonk** [comment]
@scranen do you think you can solve the merge conflict? This feature is something we're looking forward to using this, and we'd rather have this in a keycloak release than start using you branch/SNAPSHOT version :)

Anyway: thanks for implementing it!

**Comment by scranen** [comment]
@bramvonk I probably can, but this PR depends on #4356, which is blocked right now -- the RedHat folks don't like the tests, but they cannot give me any pointers on what the tests _should_ look like. If you can afford to spend some time to get this moving again, then I'm happy to help making the necessary changes to get both PRs accepted.

**Comment by stianst** [comment]
@scranen First of all thanks for your PR. We appreciate your contribution. I'm sorry it has taken so long to get this reviewed/merged and that we don't have clear answers to you on how it should be tested. We are all currently very busy with wrapping up Keycloak 3.x (RH-SSO 7.2) and I'm afraid we have to prioritise work that is required there. Spring Security adapter is not going to be supported in RH-SSO 7.2 and as such we have to wait with this a bit. Once Keycloak 3.4.2 is released (~3 weeks from now) I will make sure someone reviews this PR (and https://github.com/keycloak/keycloak/pull/4356) and either give you clear instructions on how to test or we will find someone on our end to write the tests. Again, I'm sorry about the delay here and I hope you can understand.

**Comment by trecloux** [comment]
@sebastienblanc Looks good to me

**Comment by Martin-Aulich** [comment]
Is there an estimation available when this PR will be merged & released?

**Comment by dimovelev** [comment]
any progress on this? it is a kind of blocker for using this in spring boot applications.

**Comment by stianst** [comment]
@sebastienblanc what's the status on this PR? It's been sitting around for a long time.

**Comment by bpicolo** [comment]
I, too, would appreciate this getting out

**Comment by jtorbicki** [comment]
Any news on this PR?

**Comment by pedroigor** [comment]
@jtorbicki, I'm working with @scranen in https://github.com/keycloak/keycloak/pull/4356 to get it merged first. This PR is next to be reviewed.

**Review by pedroigor — CHANGES_REQUESTED** [review]
Parent PR https://github.com/keycloak/keycloak/pull/4356 is about to be merged. Waiting Jenkis for the status.

We need to wait for parent, remove its commit from here and rebase. After that will start a final review.

**Comment by scranen** [comment]
Rebased on upstream. @pedroigor I think the Arquillian tests should be updated to test cookie storage with the Spring adapter -- could you see what's the easiest way to do that?

**Comment by yyunikov** [comment]
@pedroigor any update on this PR? would be great to see it merged

**Comment by hmlnarik** [comment]
Thank you for the contribution. Sadly, current priorities and bandwidth does not allow us to work on this PR and merge it at this moment. Therefore this PR is currently marked on hold. We will be happy to continue the work after 4.x is wrapped up.
Thank you for understanding.

**Comment by pedroigor** [comment]
@yyunikov, changes look fine. As Hynek said, we are finishing up 4.x. Once we finish, I promise it will not too long (like what happened to the other related issue :)). Thanks for understanding.

**Comment by stianst** [comment]
We're very sorry that this PR has been in the queue so long. Is there still interest in this PR? If so we can re-review and decide where to go from there?

**Comment by fluescher** [comment]
I'm still interested in this feature. Is there anything I can help with?

**Comment by SebastianSchroeder** [comment]
Yes, still interested in this PR and would be willing to help.

**Comment by stianst** [comment]
@sebastienblanc  @pedroigor  can you please review this again and provide input on what is needed to get this merged?

**Comment by pedroigor** [comment]
Changes are fine but they need some test using the quickstarts (not ideal, but what we have for integration test). I'll give it a try and see if we can merge.

**Comment by pedroigor** [comment]
I'm having issues when using the cookie store as provided by these changes. I would like to know from this group if anyone was able to get it working and possibly provide some example.

The main issue is that after successful authentication something is always forcing the adapter to redirect the user to `/sso/login` even though the adapter seems to correctly authenticate the user. I can see the `KC_REDIRECT` cookie being set and used to redirect to the URI that was saved before authenticating.

**Comment by scranen** [comment]
We're extending `KeycloakWebSecurityConfigurerAdapter` and using the following:

```java
    @Autowired
    public void configureGlobal(
            AuthenticationManagerBuilder auth, ApiKeyAuthenticationProvider apiKeyProvider)
            throws Exception {
        KeycloakAuthenticationProvider keycloakAuthenticationProvider =
                keycloakAuthenticationProvider();
        auth.authenticationProvider(keycloakAuthenticationProvider);
    }

    @Override
    protected SessionAuthenticationStrategy sessionAuthenticationStrategy() {
        return new NullAuthenticatedSessionStrategy();
    }

    @Override
    protected KeycloakAuthenticationProcessingFilter keycloakAuthenticationProcessingFilter()
            throws Exception {
        KeycloakAuthenticationProcessingFilter filter =
                super.keycloakAuthenticationProcessingFilter();
        filter.setAuthenticationSuccessHandler(KeycloakLoginRedirect::loginSuccessHandler);
        filter.setAdapterTokenStoreFactory(
                (deployment, request) ->
                        new SpringSecurityCookieTokenStore(
                                deployment,
                                request,
                                (HttpServletResponse)
                                        request.getAttribute(KeycloakCookieFilter.RESPONSE_KEY)));
        return filter;
    }

    /**
     * Prevent Spring Boot from registering filters twice.
     *
     * @see <a
     *     href="https://keycloak.gitbooks.io/documentation/securing_apps/topics/oidc/java/spring-security-adapter.html">Keycloak
     *     documentation</a>
     */
    @Bean
    public FilterRegistrationBean keycloakAuthenticationProcessingFilterRegistrationBean(
            KeycloakAuthenticationProcessingFilter filter) {
        FilterRegistrationBean registrationBean = new FilterRegistrationBean(filter);
        registrationBean.setEnabled(false);
        return registrationBean;
    }

    /**
     * Prevent Spring Boot from registering filters twice.
     *
     * @see <a
     *     href="https://keycloak.gitbooks.io/documentation/securing_apps/topics/oidc/java/spring-security-adapter.html">Keycloak
     *     documentation</a>
     */
    @Bean
    public FilterRegistrationBean keycloakPreAuthActionsFilterRegistrationBean(
            KeycloakPreAuthActionsFilter filter) {
        FilterRegistrationBean registrationBean = new FilterRegistrationBean(filter);
        registrationBean.setEnabled(false);
        return registrationBean;
    }
```

We also use a custom `KeycloakAuthenticationEntryPoint`, but I don't think that should be relevant after the authentication flow has started.

**Comment by pedroigor** [comment]
@scranen here we are again :)

So, you mean that in addition to the changes herein users should also extend the `KeycloakWebSecurityConfigurerAdapter`? And possibly the entry point?

**Comment by stianst** [comment]
Feedback from others on this PR would be very welcome.

**Comment by stianst** [comment]
I don't have any experience with Spring Security or our Spring adapter so my comments here are quite likely wrong/misguided. At this point it seems that this PR may never get merged for a few reasons:

* There is a large amount of changes to introduce support for cookies
* It seems it is rather complicated to use cookies with Spring Security and not as simple as switching to cookie store in config
* It seems no one has successfully managed to use this other than the author

The PR has been around so long without merging, so to be honest I'm not sure where to go from here.

**Comment by rainerfrey** [comment]
My company just planned to test this PR within the next two weeks, as we just decided that using cookies would be the ideal scenario for us.

The last code example from @scranen doesn't look very complicated to me, most of it is unrelated to this PR and  necessary for using the Spring Security adapter anyway.

**Comment by stianst** [comment]
@rainerfrey thanks - if you have a chance to review the changes and also test it out that would be great

**Comment by rainerfrey** [comment]
> @rainerfrey thanks - if you have a chance to review the changes and also test it out that would be great

@stianst we will definitively test it. I'm not sure in how far we can do a really qualified review as we are only users of both Keycloak and Spring Security and not that familiar with the code of either.

**Comment by scranen** [comment]
Sorry everyone, this completely fell off my radar. Our company is migrating away from using this adapter in favour of a proxy-based setup, also in part because we are migrating to Spring 5 and Spring Boot 2. @rainerfrey let me know if you have any trouble getting it to work.

**Comment by rainerfrey** [comment]
@scranen what is that `KeycloakCookieFilter` that your example refers to? I can't find it anywhere.

**Comment by rainerfrey** [comment]
> The main issue is that after successful authentication something is always forcing the adapter to redirect the user to `/sso/login` even though the adapter seems to correctly authenticate the user.

@pedroigor I saw the same behaviour, for me these two steps fixed it:
* set a`NullAuthenticatedSessionStrategy` as session authentication strategy
* set the adapter state cookie path to `/`

**Comment by rainerfrey** [comment]
@scranen please correct me if I'm wrong but regarding your example configuration above:

the custom adapter token store factory call should be unnecessary, right?
```
        filter.setAdapterTokenStoreFactory(
                (deployment, request) ->
                        new SpringSecurityCookieTokenStore(
                                deployment,
                                request,
                                (HttpServletResponse)
                                        request.getAttribute(KeycloakCookieFilter.RESPONSE_KEY)));
```
This should be handled by your change in adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/token/SpringSecurityAdapterTokenStoreFactory.java?

Also the call in the HttpSecurity setup to delete the cookie: isn't that handled by your changes to adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakLogoutHandler.java?
```           http.sessionManagement()
                .logout()
                .deleteCookies(KEYCLOAK_ADAPTER_STATE_COOKIE)
```

**Comment by rainerfrey** [comment]
As a general feedback: the code is working for me as soon as I got the setup right, but with the following caveat:
* Spring Security's session creation policy must be `STATELESS`
* as soon as something else creates a `JSESSIONID` cookie (in my case the default CSRF protection) authentication stops working - I fixed this by using the cookie based CSRF token store instead
This still works for me, as the scenario for us is to avoid using HTTP sessions.

@scranen @pedroigor is this expected and generally acceptable?
Or is there any more setup that I should do to get it to work even if a HTTP session is created?
Or does this feature need more work to make it work in that case?

I'll continue to test more scenarios during the next days, but so far this looks good to me and I'd be glad to see this in the adapter soon.

**Comment by rainerfrey** [comment]
I tested further with our actual backend application and a JavaScript SPA frontend, and once CORS and CSRF protection settings are working, everything works as expected.

From my understanding, the code looks plausible with no obvious problems. Unfortunately I am not very familiar with the internals of the Keycloak and the Spring Security codebase, so I am probably not able to catch any potentially problematic subtleties.

@scranen  @pedroigor @stianst what can I do (more than the feedback that I gave up to now) to help getting this merged? I'd be very interested to see this in a released version.

**Comment by scranen** [comment]
@rainerfrey Again sorry for the late response.

> @scranen what is that KeycloakCookieFilter that your example refers to? I can't find it anywhere.
> the custom adapter token store factory call should be unnecessary, right?
> the call in the HttpSecurity setup to delete the cookie: isn't that handled by your changes to KeycloakLogoutHandler.java?

I think you are right. The snippet I posted is the configuration we're using in production code, which is based on the current version of the adapter. Instead of maintaining a fork, we decided to just write some additional code around the vanilla adapter to plug in the various bits and pieces. IIRC, KeycloakCookieFilter is one of those bits of additional code that should be redundant if the changes are made in the adapter itself. Since it's more than 2 years ago since I wrote this I don't have everything crystal clear, so the best way to verify this is by testing that the cookie appears and disappears at the right moments.

> @scranen @pedroigor is this expected and generally acceptable?
> Or is there any more setup that I should do to get it to work even if a HTTP session is created?
> Or does this feature need more work to make it work in that case?

For us the only reason to use cookies was to be able to run our services in a stateless fashion (with the only connection state being kept in the cookie). I don't really see a use case for using both sessions _and_ cookies; if you're using sessions anyway then you might as well store your security token in the session and save yourself the trouble of dealing with the (rather large) Keycloak cookies.

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java
Comment (rainerfrey) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java]: This does not preserve query parameters of the original request. Is this intentional? Is there any (e.g. security related) reason to not include the query?

**Comment by rainerfrey** [comment]
> Since it's more than 2 years ago since I wrote this I don't have everything crystal clear, so the best way to verify this is by testing that the cookie appears and disappears at the right moments.

This is exactlyy what I have done right now.

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java
Comment (scranen) [reply adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java]: Not intenional, this is an oversight on my part. I think it would make perfect sense to include the query parameters.

**Comment by pedroigor** [comment]
@scranen You should be proud that your code is still useful two years later :)

@rainerfrey @stianst I would just merge this PR if @rainerfrey says we are good to go, even with the additions to get it working. Of course, as long as it won't break people using the adapter without this new capability.

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationSuccessHandler.java
Comment (rainerfrey) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationSuccessHandler.java]:
```suggestion
                response.addCookie(KeycloakCookieBasedRedirect.createCookieFromRedirectUrl(null));
```

otherwise the cookie is not unset but is simply set to an empty value

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java
Comment (rainerfrey) [reply adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/authentication/KeycloakAuthenticationEntryPoint.java]: > If the current Keycloak adapter already has an endpoint that finalises the authentication flow which accepts a redirect URI parameter, then maybe we could reuse that and avoid the extra cookie (which is a burden because of the HTTP headers growing rather large)

Sounds good in principle, but I am not sure if that is the case. I wouldn't worry too much about it though, because this additional cookie is so much smaller than the ADAPTER_STATE cookie. If header size is a concern, this entire approach isn't ideal.

**Comment by rainerfrey** [comment]
> I would just merge this PR if @rainerfrey says we are good to go, even with the additions to get it working.
The one change I suggested/requested in KeycloakAuthenticationSuccessHandler is fixing an actual bug and should be included - otherwise after login redirect only works once in a browsing session. The remarks in the KeycloakAuthenticationEntryPoint concern a limitation regarding query parameters in after-login redirects that doesn't concern me personally as we need to override this anyway. So a straight-forward fix would be generally desirable but not important for my use case.

With that considered, I'd say this is good to go.

The necessary changes to the application to get this working are IMO unavoidable:
* actually configure using the cookie the token store
* configure session handling and features that might trigger session creation

> Of course, as long as it won't break people using the adapter without this new capability.

I didn't test this extensively, but a cursory test without configuring the cookie store showed no change in behavior compared to the released version of the adapter. Also the maven tests ran successfully.

**Comment by cfecherolle** [comment]
Hi everyone,

I stumbled upon this PR while searching for issues regarding `tokenStore = cookie` configuration in our project, which seems not to be working for us. (cookie is never set by Keycloak even after configuring the adapter with this parameter)
Considering discussions are still ongoing but implementation seems to be ready, do you think it is going to be merged soon?

If not, is there a workaround to use the adapter in "cookie mode" instead of session with this PR? (this is a naive question, I think not, but I'd rather ask)

Thanks for your feedback.

**Comment by rainerfrey-inxmail** [comment]
> If not, is there a workaround to use the adapter in "cookie mode" instead of session without this PR? (this is a naive question, I think not, but I'd rather ask)

While I am all for merging this as soon as anyone gets to it, the easiest workaround is to build this branch yourself. (At least a few days ago) it rebased cleanly onto 7.0.0 or master. To only publish the modified Spring Security adapter, I modified the parent pom.xml, changed all the dependencies in group "org.keycloak" from $project.version" to $project.parent.version, adapted the ,even settings for our internal repository and executed "mvn deploy" in this adapter's directory.

**Review by pedroigor — DISMISSED** [review]

**Comment by pedroigor** [comment]
@rainerfrey, I've approved. Could you just provide a summary for future reference about the changes people need to do in order to use the stateless model provided by this set of changes ?

From what I've experienced, we need to:

* Enable the cookie store in adapter configuration
* Disable CSRF (or use something that does not create sessions) and any other mechanism that creates session (no JSESSIONID cookie)
* Anything else ?

**Comment by rainerfrey** [comment]
Hi @pedroigor these are the main points. Additionally:
* configure the session authentication strategy as NullAuthenticatedSessionStrategy (the docs already describe this for bearer-only applications)

**Comment by pedroigor** [comment]
@rainerfrey Thanks. I'm going to update docs with these tricks.

I also found a minor issue that impacts the policy enforcer functionality. @scranen, I can send you the patch, otherwise, I can just send another PR with your commits + the patch. Please, let me know how do you want to proceed.

Let's try to get this finally sorted out on Mon :) And thanks for your patience.

**Comment by scranen** [comment]
@pedroigor I added you and @rainerfrey as collaborators to my fork, so you should be able to apply any changes yourself now. You can also send me a patch and I'll apply it

**Comment by pedroigor** [comment]
@scranen to make things worse I messed up your branch. I'm sending another PR squashing all commits so far. Sorry for that ...

**Comment by mschielmann** [comment]
@rainerfrey on October 14th 2019 you've mentioned / asked to:
`set the adapter state cookie path to / as it defaulted to /sso`
I think this might have been missed. Currently, the cookie path is taken from the contextPath, which by default happens to be `/sso/login`, and the cookie path becomes `/sso`. 
`/sso/login` is taken from `KeycloakAuthenticationEntryPoint` where it's defaulted as:
`public static final String DEFAULT_LOGIN_URI = "/sso/login";`

Even though you may override the default, it seems that `KeycloakAuthenticationProcessingFilter` would not pick that overriden value and that is causing other problems - like the auth flow not being triggered at all.

Maybe I'm missing something, but the `/sso` path seems to be a problem for me and I've found your comment about it here, so maybe this is a bug?

**Pull request closed without merging**
